File: models/resume.go
package models

// Resume represents the JSONResume v1.0.0 schema with backward compatibility
// for legacy v0.x field names.
type Resume struct {
	Basics struct {
		Name    string `json:"name,omitempty"`
		Label   string `json:"label,omitempty"`
		Image   string `json:"image,omitempty"`   // v1.0.0
		Picture string `json:"picture,omitempty"` // legacy v0.x
		Email   string `json:"email,omitempty"`
		Phone   string `json:"phone,omitempty"`
		URL     string `json:"url,omitempty"`     // v1.0.0
		Website string `json:"website,omitempty"` // legacy v0.x
		Summary string `json:"summary,omitempty"`
		Location struct {
			Address     string `json:"address,omitempty"`
			PostalCode  string `json:"postalCode,omitempty"`
			City        string `json:"city,omitempty"`
			CountryCode string `json:"countryCode,omitempty"`
			Region      string `json:"region,omitempty"`
		} `json:"location,omitempty"`
		Profiles []struct {
			Network  string `json:"network,omitempty"`
			Username string `json:"username,omitempty"`
			URL      string `json:"url,omitempty"`
		} `json:"profiles,omitempty"`
	} `json:"basics,omitempty"`
	Work []struct {
		Name       string   `json:"name,omitempty"`    // v1.0.0
		Company    string   `json:"company,omitempty"` // legacy v0.x
		Location   string   `json:"location,omitempty"`
		Position   string   `json:"position,omitempty"`
		URL        string   `json:"url,omitempty"`     // v1.0.0
		Website    string   `json:"website,omitempty"` // legacy v0.x
		StartDate  string   `json:"startDate,omitempty"`
		EndDate    string   `json:"endDate,omitempty"`
		Summary    string   `json:"summary,omitempty"`
		Highlights []string `json:"highlights,omitempty"`
	} `json:"work,omitempty"`
	Volunteer []struct {
		Organization string   `json:"organization,omitempty"`
		Position     string   `json:"position,omitempty"`
		URL          string   `json:"url,omitempty"`     // v1.0.0
		Website      string   `json:"website,omitempty"` // legacy v0.x
		StartDate    string   `json:"startDate,omitempty"`
		EndDate      string   `json:"endDate,omitempty"`
		Summary      string   `json:"summary,omitempty"`
		Highlights   []string `json:"highlights,omitempty"`
	} `json:"volunteer,omitempty"`
	Education []struct {
		Institution string   `json:"institution,omitempty"`
		URL         string   `json:"url,omitempty"` // v1.0.0
		Location    string   `json:"location,omitempty"`
		Area        string   `json:"area,omitempty"`
		StudyType   string   `json:"studyType,omitempty"`
		StartDate   string   `json:"startDate,omitempty"`
		EndDate     string   `json:"endDate,omitempty"`
		Score       string   `json:"score,omitempty"` // v1.0.0
		GPA         string   `json:"gpa,omitempty"`   // legacy v0.x
		Courses     []string `json:"courses,omitempty"`
	} `json:"education,omitempty"`
	Awards []struct {
		Title   string `json:"title,omitempty"`
		Date    string `json:"date,omitempty"`
		Awarder string `json:"awarder,omitempty"`
		Summary string `json:"summary,omitempty"`
	} `json:"awards,omitempty"`
	Certificates []struct {
		Name   string `json:"name,omitempty"`
		Date   string `json:"date,omitempty"`
		Issuer string `json:"issuer,omitempty"`
		URL    string `json:"url,omitempty"`
	} `json:"certificates,omitempty"`
	Publications []struct {
		Name        string `json:"name,omitempty"`
		Publisher   string `json:"publisher,omitempty"`
		ReleaseDate string `json:"releaseDate,omitempty"`
		URL         string `json:"url,omitempty"`     // v1.0.0
		Website     string `json:"website,omitempty"` // legacy v0.x
		Summary     string `json:"summary,omitempty"`
	} `json:"publications,omitempty"`
	Projects []struct {
		Name        string   `json:"name,omitempty"`
		Description string   `json:"description,omitempty"` // v1.0.0
		Summary     string   `json:"summary,omitempty"`     // legacy v0.x
		Highlights  []string `json:"highlights,omitempty"`  // v1.0.0
		Keywords    []string `json:"keywords,omitempty"`    // v1.0.0
		StartDate   string   `json:"startDate,omitempty"`   // v1.0.0
		EndDate     string   `json:"endDate,omitempty"`     // v1.0.0
		URL         string   `json:"url,omitempty"`         // v1.0.0
		Website     string   `json:"website,omitempty"`     // legacy v0.x
		Roles       []string `json:"roles,omitempty"`       // v1.0.0
		Entity      string   `json:"entity,omitempty"`      // v1.0.0
		Type        string   `json:"type,omitempty"`        // v1.0.0
	} `json:"projects,omitempty"`
	Skills []struct {
		Name     string   `json:"name,omitempty"`
		Level    string   `json:"level,omitempty"`
		Keywords []string `json:"keywords,omitempty"`
	} `json:"skills,omitempty"`
	Languages []struct {
		Language string `json:"language,omitempty"`
		Fluency  string `json:"fluency,omitempty"`
	} `json:"languages,omitempty"`
	Interests []struct {
		Name     string   `json:"name,omitempty"`
		Keywords []string `json:"keywords,omitempty"`
	} `json:"interests,omitempty"`
	References []struct {
		Name      string `json:"name,omitempty"`
		Reference string `json:"reference,omitempty"`
	} `json:"references,omitempty"`
	Meta struct {
		Canonical    string `json:"canonical,omitempty"`
		Version      string `json:"version,omitempty"`
		LastModified string `json:"lastModified,omitempty"`
	} `json:"meta,omitempty"`
}

// WorkRole represents a single role/position at a company
type WorkRole struct {
	Position   string
	StartDate  string
	EndDate    string
	Summary    string
	Highlights []string
}

// GroupedWork represents a company with potentially multiple roles
type GroupedWork struct {
	Name     string
	Location string
	URL      string
	Roles    []WorkRole
}

// GroupedWorkEntries returns work entries grouped by company name.
// Consecutive entries with the same company name are merged into one entry with multiple roles.
func (r *Resume) GroupedWorkEntries() []GroupedWork {
	if len(r.Work) == 0 {
		return nil
	}

	var grouped []GroupedWork
	var current *GroupedWork

	for _, w := range r.Work {
		companyName := w.Name
		if companyName == "" {
			companyName = w.Company
		}

		role := WorkRole{
			Position:   w.Position,
			StartDate:  w.StartDate,
			EndDate:    w.EndDate,
			Summary:    w.Summary,
			Highlights: w.Highlights,
		}

		// Check if this is the same company as the previous entry
		if current != nil && current.Name == companyName {
			current.Roles = append(current.Roles, role)
		} else {
			// New company - save previous and start new
			if current != nil {
				grouped = append(grouped, *current)
			}
			current = &GroupedWork{
				Name:     companyName,
				Location: w.Location,
				URL:      w.URL,
				Roles:    []WorkRole{role},
			}
		}
	}

	// Don't forget the last one
	if current != nil {
		grouped = append(grouped, *current)
	}

	return grouped
}

// Normalize copies legacy field values to v1.0.0 fields when the new fields are empty.
// This ensures backward compatibility with older JSONResume files.
func (r *Resume) Normalize() {
	// Basics
	if r.Basics.URL == "" && r.Basics.Website != "" {
		r.Basics.URL = r.Basics.Website
	}
	if r.Basics.Image == "" && r.Basics.Picture != "" {
		r.Basics.Image = r.Basics.Picture
	}

	// Work
	for i := range r.Work {
		if r.Work[i].Name == "" && r.Work[i].Company != "" {
			r.Work[i].Name = r.Work[i].Company
		}
		if r.Work[i].URL == "" && r.Work[i].Website != "" {
			r.Work[i].URL = r.Work[i].Website
		}
	}

	// Volunteer
	for i := range r.Volunteer {
		if r.Volunteer[i].URL == "" && r.Volunteer[i].Website != "" {
			r.Volunteer[i].URL = r.Volunteer[i].Website
		}
	}

	// Education
	for i := range r.Education {
		if r.Education[i].Score == "" && r.Education[i].GPA != "" {
			r.Education[i].Score = r.Education[i].GPA
		}
	}

	// Publications
	for i := range r.Publications {
		if r.Publications[i].URL == "" && r.Publications[i].Website != "" {
			r.Publications[i].URL = r.Publications[i].Website
		}
	}

	// Projects
	for i := range r.Projects {
		if r.Projects[i].Description == "" && r.Projects[i].Summary != "" {
			r.Projects[i].Description = r.Projects[i].Summary
		}
		if r.Projects[i].URL == "" && r.Projects[i].Website != "" {
			r.Projects[i].URL = r.Projects[i].Website
		}
	}
}

File: src/resume.rs
//! JSONResume v1.0.0 data model
//! https://jsonresume.org/schema/

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Resume {
    #[serde(default)]
    pub basics: Basics,
    #[serde(default)]
    pub work: Vec<Work>,
    #[serde(default)]
    pub volunteer: Vec<Volunteer>,
    #[serde(default)]
    pub education: Vec<Education>,
    #[serde(default)]
    pub awards: Vec<Award>,
    #[serde(default)]
    pub certificates: Vec<Certificate>,
    #[serde(default)]
    pub publications: Vec<Publication>,
    #[serde(default)]
    pub skills: Vec<Skill>,
    #[serde(default)]
    pub languages: Vec<Language>,
    #[serde(default)]
    pub interests: Vec<Interest>,
    #[serde(default)]
    pub references: Vec<Reference>,
    #[serde(default)]
    pub projects: Vec<Project>,
    #[serde(default)]
    pub meta: Meta,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Basics {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub label: String,
    #[serde(default)]
    pub image: String,
    #[serde(default)]
    pub email: String,
    #[serde(default)]
    pub phone: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub summary: String,
    #[serde(default)]
    pub location: Location,
    #[serde(default)]
    pub profiles: Vec<Profile>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Location {
    #[serde(default)]
    pub address: String,
    #[serde(default)]
    pub postal_code: String,
    #[serde(default)]
    pub city: String,
    #[serde(default)]
    pub country_code: String,
    #[serde(default)]
    pub region: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Profile {
    #[serde(default)]
    pub network: String,
    #[serde(default)]
    pub username: String,
    #[serde(default)]
    pub url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Work {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub location: String,
    #[serde(default)]
    pub position: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub start_date: String,
    #[serde(default)]
    pub end_date: String,
    #[serde(default)]
    pub summary: String,
    #[serde(default)]
    pub highlights: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Volunteer {
    #[serde(default)]
    pub organization: String,
    #[serde(default)]
    pub position: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub start_date: String,
    #[serde(default)]
    pub end_date: String,
    #[serde(default)]
    pub summary: String,
    #[serde(default)]
    pub highlights: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Education {
    #[serde(default)]
    pub institution: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub area: String,
    #[serde(default)]
    pub study_type: String,
    #[serde(default)]
    pub start_date: String,
    #[serde(default)]
    pub end_date: String,
    #[serde(default)]
    pub score: String,
    #[serde(default)]
    pub location: String,
    #[serde(default)]
    pub courses: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Award {
    #[serde(default)]
    pub title: String,
    #[serde(default)]
    pub date: String,
    #[serde(default)]
    pub awarder: String,
    #[serde(default)]
    pub summary: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Certificate {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub date: String,
    #[serde(default)]
    pub issuer: String,
    #[serde(default)]
    pub url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Publication {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub publisher: String,
    #[serde(default)]
    pub release_date: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub summary: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Skill {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub level: String,
    #[serde(default)]
    pub keywords: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Language {
    #[serde(default)]
    pub language: String,
    #[serde(default)]
    pub fluency: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Interest {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub keywords: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Reference {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub reference: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Project {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub highlights: Vec<String>,
    #[serde(default)]
    pub keywords: Vec<String>,
    #[serde(default)]
    pub start_date: String,
    #[serde(default)]
    pub end_date: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub roles: Vec<String>,
    #[serde(default)]
    pub entity: String,
    #[serde(default, rename = "type")]
    pub project_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Meta {
    #[serde(default)]
    pub canonical: String,
    #[serde(default)]
    pub version: String,
    #[serde(default)]
    pub last_modified: String,
}

// Grouped work entries for templates that show multiple roles per company

#[derive(Debug, Clone, Serialize)]
pub struct WorkRole {
    pub position: String,
    pub start_date: String,
    pub end_date: String,
    pub summary: String,
    pub highlights: Vec<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct GroupedWork {
    pub name: String,
    pub location: String,
    pub url: String,
    pub roles: Vec<WorkRole>,
}

impl Resume {
    /// Group consecutive work entries by company name.
    /// Consecutive entries with the same company are merged into one with multiple roles.
    pub fn grouped_work(&self) -> Vec<GroupedWork> {
        if self.work.is_empty() {
            return Vec::new();
        }

        let mut grouped: Vec<GroupedWork> = Vec::new();
        let mut current: Option<GroupedWork> = None;

        for w in &self.work {
            let role = WorkRole {
                position: w.position.clone(),
                start_date: w.start_date.clone(),
                end_date: w.end_date.clone(),
                summary: w.summary.clone(),
                highlights: w.highlights.clone(),
            };

            if let Some(ref mut cur) = current {
                if cur.name == w.name {
                    // Same company - add role
                    cur.roles.push(role);
                } else {
                    // Different company - save current and start new
                    grouped.push(current.take().unwrap());
                    current = Some(GroupedWork {
                        name: w.name.clone(),
                        location: w.location.clone(),
                        url: w.url.clone(),
                        roles: vec![role],
                    });
                }
            } else {
                // First entry
                current = Some(GroupedWork {
                    name: w.name.clone(),
                    location: w.location.clone(),
                    url: w.url.clone(),
                    roles: vec![role],
                });
            }
        }

        // Don't forget the last one
        if let Some(cur) = current {
            grouped.push(cur);
        }

        grouped
    }
}
